### PR TITLE
script: Implement the `width` and `height` attributes for iframes per HTML5 § 4.8.6.

### PR DESCRIPTION
**Pull request by pcwalton**
Improves Amazon and Ars Technica.

r? @jdm

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/5577)

### CODE REVIEW

File: components/script/dom/htmliframeelement.rs [hunk 1]
```diff
@@ -317,6 +321,26 @@ impl<'a> HTMLIFrameElementMethods for JSRef<'a, HTMLIFrameElement> {
     fn Stop(self) -> Fallible<()> {
         Err(NotSupported)
     }
+
+    fn Width(self) -> DOMString {
+        let element: JSRef<Element> = ElementCast::from_ref(self);
```
Comment (nox) [inline components/script/dom/htmliframeelement.rs]: AFAIK the explicit types are useless now.

**Comment by jdm** [comment]
@Ms2ger I started reviewing this, but it's complicated by HTMLIFrameElement::width/height being DOMString values instead of u32 in the spec. How should we deal with this?

**Comment by Ms2ger** [comment]
The `width`/`height` attributes should simply use `make_getter!()`/`make_setter!()`. However, it looks like these attributes should yield preshints rather than being read directly from layout, as well as supporting percentages.

**Comment by pcwalton** [comment]
Rewrote the patch per @Ms2ger's review comments.

r? @jdm

**Comment by jdm** [comment]
This is missing WPT result changes.
-S-awaiting-review +S-needs-code-changes 

---

Reviewed files:
- components/layout/construct.rs @ r1, r5
- components/layout/fragment.rs @ r1, r5
- components/layout/wrapper.rs @ r2, r5
- components/script/dom/attr.rs @ r5
- components/script/dom/element.rs @ r4, r5
- components/script/dom/htmliframeelement.rs @ r5
- components/script/dom/webidls/HTMLIFrameElement.webidl @ r1
- components/style/legacy.rs @ r5
- components/util/str.rs @ r5
- tests/ref/basic.list @ r5
- tests/ref/iframe/size_attributes.html @ r1
- tests/ref/iframe/size_attributes_ref.html @ r1
- tests/ref/iframe/size_attributes_vertical_writing_mode.html @ r3

---

<sup>**[components/script/dom/htmliframeelement.rs, line 399 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5577#-Jp8un7aNm3Ix9Bp1ckv)** ([raw file](https://github.com/servo/servo/blob/68b725dd5cc496e8b583db9ef90209b4cd8ef864/components/script/dom/htmliframeelement.rs#L399)):</sup>
I don't think this is needed/desirable any longer.

---

<sup>**[components/style/legacy.rs, line 193 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5577#-Jp8tyP-1t6YoZYbzSpq)** ([raw file](https://github.com/servo/servo/blob/68b725dd5cc496e8b583db9ef90209b4cd8ef864/components/style/legacy.rs#L193)):</sup>
This code needs to move to RawLayoutElementHelpers::synthesize_presentational_hints_for_legacy_attributes now.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5577)

**Comment by pcwalton** [comment]
Rebased, updated, and comments addressed. r? @jdm

**Comment by jdm** [comment]
@bors-servo: r+
-S-awaiting-review +S-awaiting-merge 

---

Review status: :shipit: all files reviewed, all discussions resolved, all commit checks successful.
Reviewed files:
- components/layout/construct.rs @ r6
- components/layout/fragment.rs @ r6
- components/layout/wrapper.rs @ r6
- components/script/dom/attr.rs @ r6
- components/script/dom/element.rs @ r6
- components/script/dom/htmliframeelement.rs @ r6
- components/style/legacy.rs @ r6
- tests/ref/basic.list @ r6

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5577)

**Comment by jdm** [comment]
```
/html/dom/interfaces.html
-------------------------
PASS expected FAIL HTMLIFrameElement interface: attribute width
PASS expected FAIL HTMLIFrameElement interface: attribute height
/html/dom/reflection-embedded.html
----------------------------------
PASS expected FAIL iframe.width: typeof IDL attribute
PASS expected FAIL iframe.width: IDL get with DOM attribute unset
PASS expected FAIL iframe.width: setAttribute() to "" followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to " \0\x01\x02\x03\x04\x05\x06\x07 \b\t\n\v\f\r\x0e\x0f \x10\x11\x12\x13\x14\x15\x16\x17 \x18\x19\x1a\x1b\x1c\x1d\x1e\x1f  foo " followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to undefined followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to 7 followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to 1.5 followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to true followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to false followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to object "[object Object]" followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to NaN followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to Infinity followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to -Infinity followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to "\0" followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to null followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to object "test-toString" followed by IDL get
PASS expected FAIL iframe.width: setAttribute() to object "test-valueOf" followed by IDL get
PASS expected FAIL iframe.width: IDL set to "" followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to " \0\x01\x02\x03\x04\x05\x06\x07 \b\t\n\v\f\r\x0e\x0f \x10\x11\x12\x13\x14\x15\x16\x17 \x18\x19\x1a\x1b\x1c\x1d\x1e\x1f  foo " followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to undefined followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to undefined followed by IDL get
PASS expected FAIL iframe.width: IDL set to 7 followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to 7 followed by IDL get
PASS expected FAIL iframe.width: IDL set to 1.5 followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to 1.5 followed by IDL get
PASS expected FAIL iframe.width: IDL set to true followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to true followed by IDL get
PASS expected FAIL iframe.width: IDL set to false followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to false followed by IDL get
PASS expected FAIL iframe.width: IDL set to object "[object Object]" followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to object "[object Object]" followed by IDL get
PASS expected FAIL iframe.width: IDL set to NaN followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to NaN followed by IDL get
PASS expected FAIL iframe.width: IDL set to Infinity followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to Infinity followed by IDL get
PASS expected FAIL iframe.width: IDL set to -Infinity followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to -Infinity followed by IDL get
PASS expected FAIL iframe.width: IDL set to "\0" followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to null followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to null followed by IDL get
PASS expected FAIL iframe.width: IDL set to object "test-toString" followed by getAttribute()
PASS expected FAIL iframe.width: IDL set to object "test-toString" followed by IDL get
PASS expected FAIL iframe.width: IDL set to object "test-valueOf" followed by IDL get
PASS expected FAIL iframe.height: typeof IDL attribute
PASS expected FAIL iframe.height: IDL get with DOM attribute unset
PASS expected FAIL iframe.height: setAttribute() to "" followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to " \0\x01\x02\x03\x04\x05\x06\x07 \b\t\n\v\f\r\x0e\x0f \x10\x11\x12\x13\x14\x15\x16\x17 \x18\x19\x1a\x1b\x1c\x1d\x1e\x1f  foo " followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to undefined followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to 7 followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to 1.5 followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to true followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to false followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to object "[object Object]" followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to NaN followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to Infinity followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to -Infinity followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to "\0" followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to null followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to object "test-toString" followed by IDL get
PASS expected FAIL iframe.height: setAttribute() to object "test-valueOf" followed by IDL get
PASS expected FAIL iframe.height: IDL set to "" followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to " \0\x01\x02\x03\x04\x05\x06\x07 \b\t\n\v\f\r\x0e\x0f \x10\x11\x12\x13\x14\x15\x16\x17 \x18\x19\x1a\x1b\x1c\x1d\x1e\x1f  foo " followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to undefined followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to undefined followed by IDL get
PASS expected FAIL iframe.height: IDL set to 7 followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to 7 followed by IDL get
PASS expected FAIL iframe.height: IDL set to 1.5 followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to 1.5 followed by IDL get
PASS expected FAIL iframe.height: IDL set to true followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to true followed by IDL get
PASS expected FAIL iframe.height: IDL set to false followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to false followed by IDL get
PASS expected FAIL iframe.height: IDL set to object "[object Object]" followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to object "[object Object]" followed by IDL get
PASS expected FAIL iframe.height: IDL set to NaN followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to NaN followed by IDL get
PASS expected FAIL iframe.height: IDL set to Infinity followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to Infinity followed by IDL get
PASS expected FAIL iframe.height: IDL set to -Infinity followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to -Infinity followed by IDL get
PASS expected FAIL iframe.height: IDL set to "\0" followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to null followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to null followed by IDL get
PASS expected FAIL iframe.height: IDL set to object "test-toString" followed by getAttribute()
PASS expected FAIL iframe.height: IDL set to object "test-toString" followed by IDL get
PASS expected FAIL iframe.height: IDL set to object "test-valueOf" followed by IDL get
```

**Comment by mmatyas** [comment]
Does this support width/height values in percentage (eg `width="50%"`) ?

**Pull request merged**